web/shared/src/typeahead.ts

```diff
@@ -66,11 +93,32 @@ export function query_matches_string(
         return source_str.includes(query);
     }
 
-    // If there is a separator character in the query, then we
-    // require the match to start at the start of a token.
-    // (E.g. for 'ab cd ef', query could be 'ab c' or 'cd ef',
-    // but not 'b cd ef'.)
-    return source_str.startsWith(query) || source_str.includes(split_char + query);
+    if (order_matters) {
+        // If there is a separator character in the query, then we
+        // require the match to start at the start of a token.
+        // (E.g. for 'ab cd ef', query could be 'ab c' or 'cd ef',
+        // but not 'b cd ef'.)
+        return source_str.startsWith(query) || source_str.includes(split_char + query);
+    }
+    // When order does not matter, we sort words in the query and source
+    // string then try to find the best match for each query word, using
+    // `upper_bound_lexicographically`. As soon as we find a query word
+    // with no unique match, we return false, else return true at the end
+    const search_words = query.split(split_char).filter(Boolean);
+    const source_words = source_str.split(split_char).filter(Boolean);
+    if (search_words.length > source_words.length) {
+        return false;
+    }
+    search_words.sort(strcmp).reverse();
+    source_words.sort(strcmp);
+    for (const word of search_words) {
+        const match_index = upper_bound_lexicographically(word, source_words) - 1;
+        if (match_index === -1) {
+            return false;
+        }
+        source_words.splice(match_index, 1);
```

Comment:
I'm not yet convinced this has the semantics you're going for. Once a query term matches some source term as a prefix, we remove a term it matches, but there may have been several such terms. What if a later query term would have matched the one we removed, if we hadn't removed it? Can that happen? If not, this calls for some more comments explaining why not.

Reply:
A later query term won't ever match a term that's already been matched (and removed). Basically, taking query words in reverse lexicographic order, one by one, and finding for each the lexicographically last possible prefix match ensures that each query word is matched with the "best" possible source word.

I'll hopefully be able explain this better in a comment.